File: solutions/javascript/strain/1/strain.js
const partition = (list, pred) => {
  return list.reduce(([yes, no], val) => pred(val) ? [[...yes, val], no] : [yes, [...no, val]], [[], []]);
};

export const keep = (list, pred) => {
  return partition(list, pred)[0];
};

export const discard = (list, pred) => {
  return partition(list, pred)[1];
};

File: solutions/haskell/strain/1/src/Strain.hs
module Strain (keep, discard) where

discard :: (a -> Bool) -> [a] -> [a]
discard p [] = []
discard p (x : xs) = if p x then discard p xs else x : (discard p xs)

keep :: (a -> Bool) -> [a] -> [a]
keep p [] = []
keep p (x : xs) = if p x then x : (keep p xs) else keep p xs
